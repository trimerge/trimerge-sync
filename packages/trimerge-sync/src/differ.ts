import { CommitInfo } from './types';

export type ComputeRefFn<Delta, EditMetadata> = (
  baseRef: string | undefined,
  mergeRef: string | undefined,
  delta: Delta | undefined,
  editMetadata: EditMetadata,
) => string;

export type DiffFn<SavedDoc, Delta> = (
  prior: SavedDoc | undefined,
  doc: SavedDoc,
) => Delta | undefined;

export type PatchFn<SavedDoc, Delta> = (
  priorOrNext: SavedDoc | undefined,
  delta: Delta | undefined,
) => SavedDoc;

export type DocAndMetadata<Doc, EditMetadata> = {
  doc: Doc;
  metadata: EditMetadata;
};
export type CommitDoc<Doc, EditMetadata> = {
  ref: string;
} & DocAndMetadata<Doc, EditMetadata>;

export type MigrateDocFn<SavedDoc, LatestDoc extends SavedDoc, EditMetadata> = (
  doc: SavedDoc,
  editMetadata: EditMetadata,
) => DocAndMetadata<LatestDoc, EditMetadata>;

export type MergeHelpers<LatestDoc, EditMetadata> = {
  getCommitInfo(ref: string): CommitInfo;
  computeLatestDoc(ref: string): CommitDoc<LatestDoc, EditMetadata>;
  addMerge(
    doc: LatestDoc,
    editMetadata: EditMetadata,
    temp: boolean,
    leftRef: string,
    rightRef: string,
  ): string;
};
export type MergeAllBranchesFn<LatestDoc, EditMetadata> = (
  branchHeadRefs: string[],
  helpers: MergeHelpers<LatestDoc, EditMetadata>,
) => void;

export interface Differ<
  SavedDoc,
  LatestDoc extends SavedDoc,
  EditMetadata,
  Delta,
> {
  readonly migrate: MigrateDocFn<SavedDoc, LatestDoc, EditMetadata>;

  /** Calculate the ref string for a given edit */
  readonly computeRef: ComputeRefFn<Delta, EditMetadata>;

  /** Computed the difference between two Docs */
  readonly diff: DiffFn<SavedDoc, Delta>;

  /** Apply a patch from one Doc to another */
  readonly patch: PatchFn<SavedDoc, Delta>;

  /** Merge all head commits */
  readonly mergeAllBranches: MergeAllBranchesFn<LatestDoc, EditMetadata>;
}
