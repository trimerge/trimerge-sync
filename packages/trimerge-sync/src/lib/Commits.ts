import { Commit } from '../types';

export type CommitRefs = {
  ref: string;
  baseRef?: string;
  mergeRef?: string;
};

// used for getting any possible refs from a general commit, refs that do not apply will be undefined.
export const asCommitRefs = (commit: Commit<unknown, unknown>): CommitRefs =>
  commit;
