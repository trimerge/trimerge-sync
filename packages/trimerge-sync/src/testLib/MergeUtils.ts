// Basic trimerge function that merges values, strings, and objects
import {
  combineMergers,
  trimergeEquality,
  trimergeObject,
  trimergeString,
} from 'trimerge';
import { create, Delta } from 'jsondiffpatch';
import { MergeDocFn, DocAndMetadata } from '../differ';
import { produce } from 'immer';
import { computeRef as computeShaRef } from 'trimerge-sync-hash';

const trimergeObjects = combineMergers(
  trimergeEquality,
  trimergeString,
  trimergeObject,
);
export const merge: MergeDocFn<any, any> = (base, left, right) => ({
  doc: trimergeObjects(base?.doc, left.doc, right.doc),
  editMetadata: {
    ref: `(${left.ref}+${right.ref})`,
    message: `merge`,
  },
});
export const jdp = create({ textDiff: { minLength: 20 } });

export function patch<T>(base: T, delta: Delta | undefined): T {
  if (delta === undefined) {
    return base;
  }
  return produce(base, (draft) => jdp.patch(draft, delta));
}

export function diff<T>(left: T, right: T) {
  return jdp.diff(left, right);
}

// Simple no-op migration for unit tests
export function migrate<Doc, EditMetadata>(
  doc: Doc,
  editMetadata: EditMetadata,
): DocAndMetadata<Doc, EditMetadata> {
  return { doc, editMetadata };
}

export function computeRef(
  baseRef: string | undefined,
  mergeRef: string | undefined,
  delta: any,
  editMetadata: any,
): string {
  return computeShaRef(baseRef, mergeRef, delta, editMetadata).slice(0, 8);
}
