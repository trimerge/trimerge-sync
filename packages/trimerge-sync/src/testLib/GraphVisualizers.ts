import { asCommitRefs } from '../lib/Commits';
import { Commit, isMergeCommit } from '../types';

type BasicGraphItem = {
  graph: string;
  step: string;
  value: any;
};
export function getBasicGraph<CommitMetadata>(
  commits: Iterable<Commit<CommitMetadata, unknown>>,
  getEditLabel: (commit: Commit<CommitMetadata, unknown>) => string,
  getValue: (commit: Commit<CommitMetadata, unknown>) => any,
): BasicGraphItem[] {
  const result = [];
  for (const commit of commits) {
    const userId = commit.userId;
    const { ref, baseRef, mergeBaseRef, mergeRef } = asCommitRefs(commit);
    if (mergeRef) {
      result.push({
        graph: `(${baseRef} + ${mergeRef}) w/ base=${mergeBaseRef} -> ${ref}`,
        step: `User ${userId}: merge`,
        value: getValue(commit),
      });
    } else {
      result.push({
        graph: `${baseRef} -> ${ref}`,
        step: `User ${userId}: ${getEditLabel(commit)}`,
        value: getValue(commit),
      });
    }
  }
  return result;
}

export function getDotGraph<CommitMetadata>(
  commits: Iterable<Commit<CommitMetadata, unknown>>,
  getEditLabel: (commit: Commit<CommitMetadata, any>) => string,
  getValue: (commit: Commit<CommitMetadata, any>) => string,
): string {
  const lines: string[] = ['digraph {'];
  for (const commit of commits) {
    lines.push(
      `"${commit.ref}" [shape=${
        isMergeCommit(commit) ? 'rectangle' : 'ellipse'
      }, label=${JSON.stringify(getValue(commit))}]`,
    );
    if (commit.baseRef) {
      if (isMergeCommit(commit)) {
        lines.push(`"${commit.baseRef}" -> "${commit.ref}" [label=left]`);
        lines.push(
          `"${commit.mergeBaseRef}" -> "${commit.ref}" [style=dashed, label=base]`,
        );
        lines.push(`"${commit.mergeRef}" -> "${commit.ref}" [label=right]`);
      } else {
        lines.push(
          `"${commit.baseRef}" -> "${commit.ref}" [label=${JSON.stringify(
            `User ${commit.userId}: ${getEditLabel(commit)}`,
          )}]`,
        );
      }
    }
  }
  lines.push('}');
  return lines.join('\n');
}
