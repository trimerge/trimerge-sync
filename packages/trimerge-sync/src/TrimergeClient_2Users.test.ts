import { Delta } from 'jsondiffpatch';
import { TrimergeClient } from './TrimergeClient';
import { Differ } from './differ';
import { MemoryStore } from './testLib/MemoryStore';
import { computeRef, diff, merge, patch } from './testLib/MergeUtils';
import { getBasicGraph } from './testLib/GraphVisualizers';
import { ClientInfo } from './types';
import { timeout } from './lib/Timeout';

type TestEditMetadata = string;
type TestState = any;
type TestPresenceState = any;

const differ: Differ<TestState, TestEditMetadata, TestPresenceState> = {
  diff,
  patch,
  computeRef,
  merge,
};

function newStore() {
  return new MemoryStore<TestEditMetadata, Delta, TestPresenceState>();
}

function makeClient(
  userId: string,
  store: MemoryStore<TestEditMetadata, Delta, TestPresenceState>,
): TrimergeClient<TestState, TestEditMetadata, Delta, TestPresenceState> {
  return new TrimergeClient(userId, 'test', store.getLocalStore, differ);
}

function basicGraph(
  store: MemoryStore<TestEditMetadata, Delta, TestPresenceState>,
  client1: TrimergeClient<
    TestState,
    TestEditMetadata,
    Delta,
    TestPresenceState
  >,
) {
  return getBasicGraph(
    store.getCommits(),
    (commit) => commit.editMetadata,
    (commit) => client1.getCommitState(commit.ref).value,
  );
}

function sortedClients(
  client: TrimergeClient<TestState, TestEditMetadata, Delta, TestPresenceState>,
) {
  return Array.from(client.clients).sort(clientSort);
}
function clientSort(
  a: ClientInfo<TestPresenceState>,
  b: ClientInfo<TestPresenceState>,
): -1 | 1 | 0 {
  if (a.userId !== b.userId) {
    return a.userId < b.userId ? -1 : 1;
  }
  if (a.clientId !== b.clientId) {
    return a.clientId < b.clientId ? -1 : 1;
  }
  return 0;
}

describe('TrimergeClient: 2 users', () => {
  it('tracks edits', async () => {
    const store = newStore();
    const client = makeClient('a', store);

    client.updateState({}, 'initialize');
    client.updateState({ hello: 'world' }, 'add hello');
    client.updateState({ hello: 'vorld' }, 'change hello');

    expect(client.state).toEqual({ hello: 'vorld' });
  });

  it('tracks non-edits', async () => {
    const store = newStore();
    const client = makeClient('a', store);

    const onStateChange = jest.fn();
    const unsub = client.subscribeState(onStateChange);
    client.updateState(undefined, 'initialize');
    await timeout();
    client.updateState(undefined, 'initialize');
    await timeout();

    expect(onStateChange.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          undefined,
        ],
      ]
    `);
    unsub();
  });
  it('tracks presence', async () => {
    const store = newStore();
    const client = makeClient('a', store);

    const onStateChange = jest.fn();
    const unsub = client.subscribeClientList(onStateChange);
    client.updatePresence('blah');
    await timeout();

    expect(onStateChange.mock.calls.slice(-1)).toMatchInlineSnapshot(`
      Array [
        Array [
          Array [
            Object {
              "clientId": "test",
              "ref": undefined,
              "self": true,
              "state": "blah",
              "userId": "a",
            },
          ],
        ],
      ]
    `);
    unsub();
  });

  it('edit syncs across two clients', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);
    const client2 = makeClient('b', store);

    // No values
    expect(client1.state).toBe(undefined);
    expect(client2.state).toBe(undefined);

    client1.updateState({}, 'initialize');

    // Client 1 is updated, but not client2
    expect(client1.state).toEqual({});
    expect(client2.state).toBe(undefined);

    await timeout();

    expect(client1.syncStatus).toEqual({
      localRead: 'ready',
      localSave: 'ready',
      remoteConnect: 'offline',
      remoteRead: 'offline',
      remoteSave: 'saving',
    });
    expect(client2.syncStatus).toMatchInlineSnapshot(
      {
        localRead: 'ready',
        localSave: 'ready',
        remoteConnect: 'offline',
        remoteRead: 'offline',
        remoteSave: 'saving',
      },
      `
      Object {
        "localRead": "ready",
        "localSave": "ready",
        "remoteConnect": "offline",
        "remoteRead": "offline",
        "remoteSave": "saving",
      }
    `,
    );

    // Client2 is updated now
    expect(client1.state).toEqual({});
    expect(client2.state).toEqual({});
  });

  it('sends presence information correctly', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);
    const client2 = makeClient('b', store);
    const client1Sub = jest.fn();

    const client1Unsub = client1.subscribeClientList(client1Sub);

    // Initial values
    expect(client1.clients).toEqual([
      {
        userId: 'a',
        self: true,
        clientId: 'test',
      },
    ]);
    expect(client2.clients).toEqual([
      {
        userId: 'b',
        self: true,
        clientId: 'test',
      },
    ]);

    expect(client1Sub.mock.calls).toEqual([
      [
        [
          {
            userId: 'a',
            self: true,
            clientId: 'test',
          },
        ],
      ],
    ]);

    await timeout();

    // Client2 is updated now
    expect(sortedClients(client1)).toEqual([
      {
        clientId: 'test',
        ref: undefined,
        self: true,
        state: undefined,
        userId: 'a',
      },
      {
        clientId: 'test',
        ref: undefined,
        state: undefined,
        userId: 'b',
      },
    ]);
    expect(sortedClients(client2)).toEqual([
      {
        clientId: 'test',
        ref: undefined,
        state: undefined,
        userId: 'a',
      },
      {
        clientId: 'test',
        ref: undefined,
        self: true,
        state: undefined,
        userId: 'b',
      },
    ]);

    expect(client1Sub.mock.calls).toEqual([
      [
        [
          {
            clientId: 'test',
            ref: undefined,
            self: true,
            state: undefined,
            userId: 'a',
          },
        ],
      ],
      [
        [
          {
            clientId: 'test',
            ref: undefined,
            self: true,
            state: undefined,
            userId: 'a',
          },
          {
            clientId: 'test',
            ref: undefined,
            state: undefined,
            userId: 'b',
          },
        ],
      ],
    ]);
    client1Unsub();
  });

  it('handles client-leave', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);
    const client2 = makeClient('b', store);

    await timeout();

    expect(sortedClients(client2)).toEqual([
      {
        clientId: 'test',
        ref: undefined,
        state: undefined,
        userId: 'a',
      },
      {
        clientId: 'test',
        ref: undefined,
        self: true,
        state: undefined,
        userId: 'b',
      },
    ]);

    await client1.shutdown();

    expect(sortedClients(client2)).toEqual([
      {
        clientId: 'test',
        ref: undefined,
        self: true,
        state: undefined,
        userId: 'b',
      },
    ]);
  });

  it('updates presence information', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);
    const client2 = makeClient('b', store);
    // Initial values
    expect(client1.clients).toEqual([
      {
        userId: 'a',
        self: true,
        clientId: 'test',
      },
    ]);
    expect(client2.clients).toEqual([
      {
        userId: 'b',
        self: true,
        clientId: 'test',
      },
    ]);

    client1.updatePresence('hello');

    await timeout();

    expect(sortedClients(client1)).toEqual([
      {
        clientId: 'test',
        ref: undefined,
        self: true,
        state: 'hello',
        userId: 'a',
      },
      {
        clientId: 'test',
        ref: undefined,
        state: undefined,
        userId: 'b',
      },
    ]);
    expect(sortedClients(client2)).toEqual([
      {
        clientId: 'test',
        ref: undefined,
        state: 'hello',
        userId: 'a',
      },
      {
        clientId: 'test',
        ref: undefined,
        self: true,
        state: undefined,
        userId: 'b',
      },
    ]);
  });

  it('two edits sync across two clients', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);
    const client2 = makeClient('b', store);

    client1.updateState({}, 'initialize');
    client1.updateState({ edit: true }, 'edit');

    // Client 1 is updated, but not client2
    expect(client1.state).toEqual({ edit: true });
    expect(client2.state).toBe(undefined);

    await timeout();

    expect(client2.state).toEqual({ edit: true });
  });

  it('edit syncs back and forth with two clients', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);
    const client2 = makeClient('b', store);

    client1.updateState({}, 'initialize');
    client1.updateState({ hello: 'world' }, 'add hello');
    client1.updateState({ hello: 'vorld' }, 'change hello');

    await timeout();

    expect(client1.state).toEqual({ hello: 'vorld' });
    expect(client2.state).toEqual({ hello: 'vorld' });

    client2.updateState({ hello: 'vorld', world: 'world' }, 'add world');
    client2.updateState({ hello: 'vorld', world: 'vorld' }, 'change world');

    // Now client 2 is updated but not client 1
    expect(client1.state).toEqual({ hello: 'vorld' });
    expect(client2.state).toEqual({ hello: 'vorld', world: 'vorld' });

    await timeout();

    expect(client1.state).toEqual({ hello: 'vorld', world: 'vorld' });
    expect(client2.state).toEqual({ hello: 'vorld', world: 'vorld' });
  });

  it('automatic merging if two clients edit simultaneously', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);
    const client2 = makeClient('b', store);

    client1.updateState({}, 'initialize');

    // Synchronized
    expect(client1.state).toEqual({});
    expect(client2.state).toEqual(undefined);

    await timeout();

    expect(client2.state).toEqual({});

    client1.updateState({ hello: 'world' }, 'add hello');
    client1.updateState({ hello: 'vorld' }, 'change hello');

    client2.updateState({ world: 'world' }, 'add world');
    client2.updateState({ world: 'vorld' }, 'change world');

    // Now client 1 and client 2 have different changes
    expect(client1.state).toEqual({ hello: 'vorld' });
    expect(client2.state).toEqual({ world: 'vorld' });

    await timeout();

    //  Now they should both have trimerged changes
    expect(client1.state).toEqual({ hello: 'vorld', world: 'vorld' });
    expect(client2.state).toEqual({ hello: 'vorld', world: 'vorld' });

    await timeout();

    // Should be the same
    expect(client1.state).toEqual({ hello: 'vorld', world: 'vorld' });
    expect(client2.state).toEqual({ hello: 'vorld', world: 'vorld' });

    await client1.shutdown();
    await client2.shutdown();

    expect(basicGraph(store, client1)).toMatchInlineSnapshot(`
      Array [
        Object {
          "graph": "undefined -> DuQe--Vh",
          "step": "User a: initialize",
          "value": Object {},
        },
        Object {
          "graph": "DuQe--Vh -> u0wBto6f",
          "step": "User a: add hello",
          "value": Object {
            "hello": "world",
          },
        },
        Object {
          "graph": "u0wBto6f -> YYUSBDXS",
          "step": "User a: change hello",
          "value": Object {
            "hello": "vorld",
          },
        },
        Object {
          "graph": "DuQe--Vh -> SU0_JahJ",
          "step": "User b: add world",
          "value": Object {
            "world": "world",
          },
        },
        Object {
          "graph": "SU0_JahJ -> ZiYUF2m8",
          "step": "User b: change world",
          "value": Object {
            "world": "vorld",
          },
        },
        Object {
          "graph": "(YYUSBDXS + ZiYUF2m8) w/ base=DuQe--Vh -> kT9Dv92V",
          "step": "User b: merge",
          "value": Object {
            "hello": "vorld",
            "world": "vorld",
          },
        },
      ]
    `);
  });

  it('sync up when second client comes in later', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);

    client1.updateState({}, 'initialize');
    client1.updateState({ hello: 'world' }, 'add hello');
    client1.updateState({ hello: 'vorld' }, 'change hello');

    await timeout();

    const client2 = makeClient('b', store);

    expect(client1.state).toEqual({ hello: 'vorld' });
    expect(client2.state).toEqual(undefined);
    await timeout();
    expect(client2.state).toEqual({ hello: 'vorld' });

    expect(basicGraph(store, client1)).toMatchInlineSnapshot(`
      Array [
        Object {
          "graph": "undefined -> DuQe--Vh",
          "step": "User a: initialize",
          "value": Object {},
        },
        Object {
          "graph": "DuQe--Vh -> u0wBto6f",
          "step": "User a: add hello",
          "value": Object {
            "hello": "world",
          },
        },
        Object {
          "graph": "u0wBto6f -> YYUSBDXS",
          "step": "User a: change hello",
          "value": Object {
            "hello": "vorld",
          },
        },
      ]
    `);
  });

  it('subscription works', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);
    const subscribeFn = jest.fn();

    const unsubscribeFn = client1.subscribeState(subscribeFn);

    client1.updateState({}, 'initialize');
    client1.updateState({ hello: 'world' }, 'add hello');
    client1.updateState({ hello: 'vorld' }, 'change hello');

    await timeout();

    const client2 = makeClient('b', store);

    expect(client1.state).toEqual({ hello: 'vorld' });
    expect(client2.state).toEqual(undefined);

    await timeout();

    client1.updateState({ hello: 'there' }, 'change hello again');

    await timeout();

    unsubscribeFn();

    expect(basicGraph(store, client1)).toMatchInlineSnapshot(`
      Array [
        Object {
          "graph": "undefined -> DuQe--Vh",
          "step": "User a: initialize",
          "value": Object {},
        },
        Object {
          "graph": "DuQe--Vh -> u0wBto6f",
          "step": "User a: add hello",
          "value": Object {
            "hello": "world",
          },
        },
        Object {
          "graph": "u0wBto6f -> YYUSBDXS",
          "step": "User a: change hello",
          "value": Object {
            "hello": "vorld",
          },
        },
        Object {
          "graph": "YYUSBDXS -> iSZbPHuf",
          "step": "User a: change hello again",
          "value": Object {
            "hello": "there",
          },
        },
      ]
    `);

    expect(subscribeFn.mock.calls).toEqual([
      [undefined],
      [{}],
      [{ hello: 'world' }],
      [{ hello: 'vorld' }],
      [{ hello: 'there' }],
    ]);
  });

  it('works with lots of character typing', async () => {
    const store = newStore();
    const client1 = makeClient('a', store);

    client1.updateState({}, 'initialize');
    client1.updateState({ hello: 'world' }, 'add hello');
    client1.updateState({ hello: 'world. t' }, 'typing');
    client1.updateState({ hello: 'world. th' }, 'typing');
    client1.updateState({ hello: 'world. thi' }, 'typing');
    client1.updateState({ hello: 'world. this' }, 'typing');
    client1.updateState({ hello: 'world. this ' }, 'typing');
    client1.updateState({ hello: 'world. this i' }, 'typing');
    client1.updateState({ hello: 'world. this is' }, 'typing');
    client1.updateState({ hello: 'world. this is ' }, 'typing');
    client1.updateState({ hello: 'world. this is a' }, 'typing');
    client1.updateState({ hello: 'world. this is a t' }, 'typing');
    client1.updateState({ hello: 'world. this is a te' }, 'typing');
    client1.updateState({ hello: 'world. this is a tes' }, 'typing');
    client1.updateState({ hello: 'world. this is a test' }, 'typing');
    client1.updateState({ hello: 'world. this is a test ' }, 'typing');
    client1.updateState({ hello: 'world. this is a test o' }, 'typing');
    client1.updateState({ hello: 'world. this is a test of' }, 'typing');
    client1.updateState({ hello: 'world. this is a test of ' }, 'typing');
    client1.updateState({ hello: 'world. this is a test of c' }, 'typing');
    client1.updateState({ hello: 'world. this is a test of ch' }, 'typing');
    client1.updateState({ hello: 'world. this is a test of cha' }, 'typing');
    client1.updateState({ hello: 'world. this is a test of char' }, 'typing');
    client1.updateState({ hello: 'world. this is a test of chara' }, 'typing');
    client1.updateState({ hello: 'world. this is a test of charac' }, 'typing');
    client1.updateState(
      { hello: 'world. this is a test of charact' },
      'typing',
    );
    client1.updateState(
      { hello: 'world. this is a test of characte' },
      'typing',
    );
    client1.updateState(
      { hello: 'world. this is a test of character' },
      'typing',
    );
    client1.updateState(
      { hello: 'world. this is a test of character.' },
      'typing',
    );

    await timeout();

    expect(basicGraph(store, client1)).toMatchInlineSnapshot(`
      Array [
        Object {
          "graph": "undefined -> DuQe--Vh",
          "step": "User a: initialize",
          "value": Object {},
        },
        Object {
          "graph": "DuQe--Vh -> u0wBto6f",
          "step": "User a: add hello",
          "value": Object {
            "hello": "world",
          },
        },
        Object {
          "graph": "u0wBto6f -> VS2jghNi",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. t",
          },
        },
        Object {
          "graph": "VS2jghNi -> 0fvI2ESx",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. th",
          },
        },
        Object {
          "graph": "0fvI2ESx -> Jz2-R6rz",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. thi",
          },
        },
        Object {
          "graph": "Jz2-R6rz -> -bTKiTst",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this",
          },
        },
        Object {
          "graph": "-bTKiTst -> GaWb8t2f",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this ",
          },
        },
        Object {
          "graph": "GaWb8t2f -> 9J_xBqJ4",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this i",
          },
        },
        Object {
          "graph": "9J_xBqJ4 -> SYH3X4jm",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is",
          },
        },
        Object {
          "graph": "SYH3X4jm -> Nl0PNGuX",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is ",
          },
        },
        Object {
          "graph": "Nl0PNGuX -> vlZhl6Vh",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a",
          },
        },
        Object {
          "graph": "vlZhl6Vh -> YxT5Gm6R",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a t",
          },
        },
        Object {
          "graph": "YxT5Gm6R -> F0g2iLQv",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a te",
          },
        },
        Object {
          "graph": "F0g2iLQv -> vKey2nks",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a tes",
          },
        },
        Object {
          "graph": "vKey2nks -> C9Ub6hg6",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test",
          },
        },
        Object {
          "graph": "C9Ub6hg6 -> ObcwRxBk",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test ",
          },
        },
        Object {
          "graph": "ObcwRxBk -> Y8mdIN_L",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test o",
          },
        },
        Object {
          "graph": "Y8mdIN_L -> L000b_2W",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of",
          },
        },
        Object {
          "graph": "L000b_2W -> tYpujZ6D",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of ",
          },
        },
        Object {
          "graph": "tYpujZ6D -> -SKX2OVN",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of c",
          },
        },
        Object {
          "graph": "-SKX2OVN -> ffe6ZCmD",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of ch",
          },
        },
        Object {
          "graph": "ffe6ZCmD -> 8PtNr3hx",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of cha",
          },
        },
        Object {
          "graph": "8PtNr3hx -> oLFFrO2p",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of char",
          },
        },
        Object {
          "graph": "oLFFrO2p -> 8Xt-akSw",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of chara",
          },
        },
        Object {
          "graph": "8Xt-akSw -> UU5J3Qq2",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of charac",
          },
        },
        Object {
          "graph": "UU5J3Qq2 -> Oo2NTgQE",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of charact",
          },
        },
        Object {
          "graph": "Oo2NTgQE -> ci7d46HK",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of characte",
          },
        },
        Object {
          "graph": "ci7d46HK -> 7_amDpeg",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of character",
          },
        },
        Object {
          "graph": "7_amDpeg -> dKj2TVjC",
          "step": "User a: typing",
          "value": Object {
            "hello": "world. this is a test of character.",
          },
        },
      ]
    `);
  });
});
