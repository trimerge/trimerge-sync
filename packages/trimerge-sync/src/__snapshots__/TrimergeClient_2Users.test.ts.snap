// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrimergeClient: 2 users it still generates merge commits even when deltas are noop 1`] = `
[
  {
    "baseRef": undefined,
    "delta": [
      {
        "hi": "world",
        "other": "world",
      },
    ],
    "metadata": {
      "message": "blah",
    },
    "ref": "83C7ugjw",
  },
  {
    "baseRef": undefined,
    "delta": [
      {
        "hi": "world",
      },
    ],
    "metadata": {
      "message": "blah",
    },
    "ref": "guqurHSi",
  },
  {
    "baseRef": "83C7ugjw",
    "delta": undefined,
    "mergeRef": "guqurHSi",
    "metadata": {
      "message": "merge",
      "ref": "(83C7ugjw+guqurHSi)",
    },
    "ref": "jBSIed8b",
  },
  {
    "baseRef": "jBSIed8b",
    "delta": {
      "other": [
        "world",
        "worldly",
      ],
    },
    "metadata": {
      "message": "blah",
    },
    "ref": "FM-iAbj3",
  },
]
`;

exports[`TrimergeClient: 2 users works with lots of character typing 1`] = `
[
  {
    "graph": "undefined -> Zob0dMmD",
    "step": "initialize",
    "value": {},
  },
  {
    "graph": "Zob0dMmD -> leySPlIR",
    "step": "add hello",
    "value": {
      "hello": "world",
    },
  },
  {
    "graph": "leySPlIR -> YAy0M_J2",
    "step": "typing",
    "value": {
      "hello": "world. t",
    },
  },
  {
    "graph": "YAy0M_J2 -> LsIxqujJ",
    "step": "typing",
    "value": {
      "hello": "world. th",
    },
  },
  {
    "graph": "LsIxqujJ -> yoPegGx6",
    "step": "typing",
    "value": {
      "hello": "world. thi",
    },
  },
  {
    "graph": "yoPegGx6 -> eTLOHYa-",
    "step": "typing",
    "value": {
      "hello": "world. this",
    },
  },
  {
    "graph": "eTLOHYa- -> WDzPFBwe",
    "step": "typing",
    "value": {
      "hello": "world. this ",
    },
  },
  {
    "graph": "WDzPFBwe -> YoyNjiZ6",
    "step": "typing",
    "value": {
      "hello": "world. this i",
    },
  },
  {
    "graph": "YoyNjiZ6 -> rOUBm7c2",
    "step": "typing",
    "value": {
      "hello": "world. this is",
    },
  },
  {
    "graph": "rOUBm7c2 -> MsplY0xo",
    "step": "typing",
    "value": {
      "hello": "world. this is ",
    },
  },
  {
    "graph": "MsplY0xo -> JnbUEhpb",
    "step": "typing",
    "value": {
      "hello": "world. this is a",
    },
  },
  {
    "graph": "JnbUEhpb -> POK9sZXI",
    "step": "typing",
    "value": {
      "hello": "world. this is a t",
    },
  },
  {
    "graph": "POK9sZXI -> yEO3XYgv",
    "step": "typing",
    "value": {
      "hello": "world. this is a te",
    },
  },
  {
    "graph": "yEO3XYgv -> VIhrAVTG",
    "step": "typing",
    "value": {
      "hello": "world. this is a tes",
    },
  },
  {
    "graph": "VIhrAVTG -> 9HSyQTMd",
    "step": "typing",
    "value": {
      "hello": "world. this is a test",
    },
  },
  {
    "graph": "9HSyQTMd -> xGjtRo_F",
    "step": "typing",
    "value": {
      "hello": "world. this is a test ",
    },
  },
  {
    "graph": "xGjtRo_F -> GFUqLq42",
    "step": "typing",
    "value": {
      "hello": "world. this is a test o",
    },
  },
  {
    "graph": "GFUqLq42 -> 8Zpd5VpF",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of",
    },
  },
  {
    "graph": "8Zpd5VpF -> 0JCZFqxq",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of ",
    },
  },
  {
    "graph": "0JCZFqxq -> 5Y4GtM8z",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of c",
    },
  },
  {
    "graph": "5Y4GtM8z -> W-adW2a-",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of ch",
    },
  },
  {
    "graph": "W-adW2a- -> 1nf6gXl1",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of cha",
    },
  },
  {
    "graph": "1nf6gXl1 -> xF9W97WS",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of char",
    },
  },
  {
    "graph": "xF9W97WS -> E8TIq05x",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of chara",
    },
  },
  {
    "graph": "E8TIq05x -> 3hCls5tY",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of charac",
    },
  },
  {
    "graph": "3hCls5tY -> Hl2TeGle",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of charact",
    },
  },
  {
    "graph": "Hl2TeGle -> NXbxkJK2",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of characte",
    },
  },
  {
    "graph": "NXbxkJK2 -> Uc41cdXS",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of character",
    },
  },
  {
    "graph": "Uc41cdXS -> QNhbrQRR",
    "step": "typing",
    "value": {
      "hello": "world. this is a test of character.",
    },
  },
]
`;
